packages/core/src/middleware/koa-consent-guard.ts

```diff
@@ -0,0 +1,71 @@
+import { experience } from '@logto/schemas';
+import { type MiddlewareType } from 'koa';
+import { type IRouterParamContext } from 'koa-router';
+import { errors, type Provider } from 'oidc-provider';
+
+import RequestError from '../errors/RequestError/index.js';
+import type Libraries from '../tenants/Libraries.js';
+import type Queries from '../tenants/Queries.js';
+import assertThat from '../utils/assert-that.js';
+
+/**
+ * Guard before allowing auto-consent
+ */
+export default function koaConsentGuard<
+  StateT,
+  ContextT extends IRouterParamContext,
+  ResponseBodyT,
+>(
+  provider: Provider,
+  libraries: Libraries,
+  query: Queries
+): MiddlewareType<StateT, ContextT, ResponseBodyT> {
+  return async (ctx, next) => {
+    const interactionDetails = await provider.interactionDetails(ctx.req, ctx.res);
+    const {
+      params: { token, login_hint: loginHint },
+      session,
+    } = interactionDetails;
+
+    assertThat(session, new errors.SessionNotFound('session not found'));
+
+    if (token && loginHint && typeof token === 'string' && typeof loginHint === 'string') {
+      const user = await query.users.findUserById(session.accountId);
+
+      assertThat(user.primaryEmail, 'user.email_not_exist');
+
+      if (user.primaryEmail !== loginHint) {
+        const searchParams = new URLSearchParams({
+          account: user.primaryEmail,
+        });
+        ctx.redirect(`${experience.routes.switchAccount}?${searchParams.toString()}`);
+        return;
+      }
+
+      try {
+        await libraries.oneTimeTokens.verifyOneTimeToken(token, loginHint);
+      } catch (error: unknown) {
+        if (error instanceof RequestError) {
+          if (error.code === 'one_time_token.email_mismatch') {
```

Comment:
Should reject the sign-in request when `login_hint` does not match with the value stored in token

Reply:
Code updated. The guard should only check if the `login_hint` matches the session user email, and does NOT need to verify the token. Token verification should only be handled in the Experience API